### DISCUSSION TITLE
How to run another simulation as a prediction during each step of the simulation?

### DISCUSSION THREAD

**Discussion by lukc19** [pyswmm/pyswmm#423] [Q&A] [answered]
Codes are as follows:
The following issue occurred during runtime:
CalledProcessError
Command '['D:/Anaconda3/envs/lkc/python.exe', 'D:/pyCode/yard_swmm_rl/swmm_run_mpc/run_swmm_fcst.py']' returned non-zero exit status 3221225477.
  File "D:\pyCode\yard_swmm_rl\swmm_run_mpc\main.py", line 76, in <module>
    fcst_submodel = subprocess.check_output("D:/Anaconda3/envs/lkc/python.exe D:/pyCode/yard_swmm_rl/swmm_run_mpc/run_swmm_fcst.py".split())

    with Simulation(tmp_file) as sim:
    sim.step_advance(control_time_step)
    node_object = Nodes(sim) #节点
    l1 = node_object["SJ1"]
    l2 = node_object["SJ2"]
    l3 = node_object["SJ3"]
    l4 = node_object["SJ4"]
    l5 = node_object["SJ5"]

    link_object = Links(sim) #管道中的阀门
    v1 = link_object["OR1"]
    v2 = link_object["OR2"]
    v3 = link_object["OR3"]
    v4 = link_object["OR4"]
        
    v1.target_setting = 1 #初始化阀门状态：全开
    v2.target_setting = 1
    v3.target_setting = 1
    v4.target_setting = 1

    current_step = 0

   
    for step in sim:
        current_dt = sim.current_time
        init_df = pd.DataFrame([l1.depth, l2.depth, l3.depth, l4.depth, l5.depth, 
                                v1.current_setting, v2.current_setting, v3.current_setting, v4.current_setting, 
                                            current_dt]).transpose()
        init_df.to_csv(os.path.join(url, "temp_model/temp_inp.csv"), index=False)
        temp_file = swmm_utils.write_ctl_dates(tmp_file, os.path.join(url, "temp_model"), current_dt)
        fcst_submodel = subprocess.check_output("D:/Anaconda3/envs/lkc/python.exe D:/pyCode/yard_swmm_rl/swmm_run_mpc/run_swmm_fcst.py".split())
        submodel_return = fcst_submodel.decode('utf-8')
        flood_dict = json.loads(submodel_return)
        print(flood_dict)

        if flood_dict["St1"] > 0:
            v2.target_setting = 0
            v3.target_setting = 0
[swmm_run_mpc.zip](https://github.com/OpenWaterAnalytics/pyswmm/files/11202774/swmm_run_mpc.zip)

**Comment by bemcdonnell** [comment]
@lukc19 are you seeing an RPT/OUT file created?

**Reply by lukc19** [reply]
Yes, when I ran it the today, I saw that there were output files, which indicated that the nested SWMM model could run! Thanks~

**Comment by karosc** [comment, marked as answer]
I was able to get your code to run in python 3.10. With some minor changes to main,py and fixing the path notation everywhere to use relative paths with double backslashes.

[swmm_run_mpc_ckEDIT.zip](https://github.com/OpenWaterAnalytics/pyswmm/files/11226075/swmm_run_mpc_ckEDIT.zip)

**Reply by lukc19** [reply]
However, when I ran into the line 'flood_dict = json.loads(submodel_return)', the level information could not be transmitted back to the current simulation properly, and the error produced was 'JSONDecodeError Expecting value: line 2 column 2 (char 2) File "D:\pyCode\yard_swmm_rl\swmm_run_mpc\main.py", line 78, in <module> flood_dict = json.loads(submodel_return)'”

**Reply by lukc19** [reply]
I was able to run your file successfully... Let me double-check where the problem might be.
